packages/core/core/src/TargetResolver.js

```diff
@@ -255,25 +261,36 @@ export default class TargetResolver {
 
     for (let targetName of COMMON_TARGETS) {
       let targetDist;
+      let pointer;
       if (
         targetName === 'browser' &&
         pkg[targetName] != null &&
         typeof pkg[targetName] === 'object'
       ) {
         // The `browser` field can be a file path or an alias map.
         targetDist = pkg[targetName][pkg.name];
+        pointer = `/${targetName}/${pkg.name}`;
       } else {
         targetDist = pkg[targetName];
+        pointer = `/${targetName}`;
       }
 
       if (typeof targetDist === 'string' || pkgTargets[targetName]) {
         let distDir;
         let distEntry;
+        let loc;
+
+        invariant(typeof pkgFilePath === 'string');
+        invariant(pkgMap != null);
 
         let _descriptor: mixed = pkgTargets[targetName] || {};
         if (typeof targetDist === 'string') {
           distDir = path.resolve(pkgDir, path.dirname(targetDist));
           distEntry = path.basename(targetDist);
+          loc = {
+            filePath: pkgFilePath,
+            ...getJSONSourceLocation(pkgMap.pointers[pointer], 'value')
+          };
         } else {
           distDir = path.resolve(pkgDir, DEFAULT_DIST_DIRNAME, targetName);
```

Comment:
Slightly off topic question: should we do this or should targets always require an explicit top-level field to define the output path? Currently you can still get the duplicate bundle name error if you had the following config:

```json
{
  "main": "dist/default/index.js",
  "targets": {
    "default": {
      "publicUrl": "/"
    }
  }
}
```

For libraries as well, if we want to allow resolvers to read the targets and use their outputs, we probably don't want to allow implicit dist paths like this...

Reply:
In the case where a `main` top-level field is defined, as is a `main` target (for listing e.g. engines or publicUrl), the `distDir` is already implicit in the top-level `main`, and could even contradict it (which I suppose is a problem on its own) if it doesn't reiterate what's in the top level field. In cases like these it could feel pretty redundant.

Maybe a better solution might be to provide a diagnostic when `distDir`s overlap for targets?

Reply:
You could output multiple targets into the same dir though...

Also, I don't think we currently support `distDir` as an option in targets (only in the programmatic API). So this would only happen if the dist was completely unspecified.